src/action/list_directory.rs

```diff
@@ -0,0 +1,689 @@
+// Copyright 2020 Google LLC
+//
+// Use of this source code is governed by an MIT-style license that can be found
+// in the LICENSE file or at https://opensource.org/licenses/MIT.
+
+//! A handler and associated types for the list directory action.
+//!
+//! A list directory action stats all files in the provided directory.
+
+use crate::session::{self, Session};
+use rrg_proto::{ListDirRequest, StatEntry, path_spec::PathType,
+                path_spec::Options};
+
+use std::fs::{self, Metadata};
+use std::path::{PathBuf, Path};
+use std::fmt::{Display, Formatter};
+use log::warn;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+#[cfg(target_os = "linux")]
+use std::os::raw::c_long;
+
+#[derive(Debug)]
+enum Error {
+    ReadPath(std::io::Error),
+}
+
+impl std::error::Error for Error {
+
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use Error::*;
+
+        match *self {
+            ReadPath(ref error) => Some(error),
+        }
+    }
+}
+
+impl Display for Error {
+
+    fn fmt(&self, fmt: &mut Formatter) -> std::fmt::Result {
+        use Error::*;
+
+        match *self {
+            ReadPath(ref error) => {
+                write!(fmt, "unable to read path: {}", error)
+            }
+        }
+    }
+}
+
+impl From<Error> for session::Error {
+
+    fn from(error: Error) -> session::Error {
+        session::Error::action(error)
+    }
+}
+
+#[derive(Debug)]
+struct UnsupportedValueError {
+    field: String,
+    value: String,
+}
+
+fn unsupported_value(field: String, value: String) -> session::ParseError {
+    session::ParseError::malformed(ParseError::UnsupportedValue(
+        UnsupportedValueError {
+            field,
+            value,
+        }))
+}
+
+#[derive(Debug)]
+enum ParseError {
+    UnsupportedValue(UnsupportedValueError),
+}
+
+impl std::error::Error for ParseError {
+
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use ParseError::*;
+
+        match *self {
+            UnsupportedValue(_) => None,
+        }
+    }
+}
+
+impl Display for ParseError {
+
+    fn fmt(&self, fmt: &mut Formatter) -> std::fmt::Result {
+        use ParseError::*;
+
+        match *self {
+            UnsupportedValue(ref message) => {
+                write!(fmt, "value {} in {} field is not supported",
+                       message.value, message.field)
+            }
+        }
+    }
+}
+
+impl From<ParseError> for session::Error {
+
+    fn from(error: ParseError) -> session::Error {
+        session::Error::action(error)
+    }
+}
+
+
+/// A response type for the list directory action.
+pub struct Response {
+    mode: Option<u64>,
+    ino: Option<u32>,
+    dev: Option<u32>,
+    nlink: Option<u32>,
+    uid: Option<u32>,
+    gid: Option<u32>,
+    size: Option<u64>,
+    atime: Option<SystemTime>,
+    mtime: Option<SystemTime>,
+    ctime: Option<SystemTime>,
```

Comment:
Some of these fields are always `Some`. No need to make them `Optional` then.

Reply:
Only `size`, I think.